x/ibc-rate-limit/contracts/rate-limiter/src/contract.rs

```diff
@@ -0,0 +1,520 @@
+    IBCMODULE.save(deps.storage, &msg.ibc_module)?;
+    GOVMODULE.save(deps.storage, &msg.gov_module)?;
+
+    add_new_channels(deps, msg.channels, env.block.time)?;
+
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
+        .add_attribute("ibc_module", msg.ibc_module.to_string())
+        .add_attribute("gov_module", msg.gov_module.to_string()))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::SendPacket {
+            channel_id,
+            channel_value,
+            funds,
+        } => try_transfer(
+            deps,
+            info.sender,
+            channel_id,
+            channel_value,
+            funds,
+            FlowType::Out,
+            env.block.time,
+        ),
+        ExecuteMsg::RecvPacket {
+            channel_id,
+            channel_value,
+            funds,
+        } => try_transfer(
+            deps,
+            info.sender,
+            channel_id,
+            channel_value,
+            funds,
+            FlowType::In,
+            env.block.time,
+        ),
+        ExecuteMsg::AddChannel { channel_id, quotas } => {
+            try_add_channel(deps, info.sender, channel_id, quotas, env.block.time)
+        }
+        ExecuteMsg::RemoveChannel { channel_id } => {
+            try_remove_channel(deps, info.sender, channel_id)
+        }
+        ExecuteMsg::ResetChannelQuota {
+            channel_id,
+            quota_id,
+        } => try_reset_channel_quota(deps, info.sender, channel_id, quota_id, env.block.time),
+    }
+}
+
+pub struct ChannelFlowResponse {
+    pub channel_flow: ChannelFlow,
+    pub used: u128,
+    pub max: u128,
+}
+
+// TODO: Lets add some more docs for this, namely that channel_value is the total supply of the denom
+// Q: Is an ICS 20 transfer only 1 denom at a time, or does the caller have to split into several
+// calls if its a multi-denom ICS-20 transfer
+pub fn try_transfer(
+    deps: DepsMut,
+    sender: Addr,
+    channel_id: String,
+    channel_value: u128,
+    funds: u128,
+    direction: FlowType,
+    now: Timestamp,
+) -> Result<Response, ContractError> {
+    // Only the IBCMODULE can execute transfers
+    // TODO: Should we move this to a helper method?
+    let ibc_module = IBCMODULE.load(deps.storage)?;
+    if sender != ibc_module {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let channels = CHANNEL_FLOWS.may_load(deps.storage, &channel_id)?;
+
+    let configured = match channels {
+        None => false,
+        Some(ref x) if x.is_empty() => false,
+        _ => true,
+    };
+
+    if !configured {
+        // No Quota configured for the current channel. Allowing all messages.
+        // TODO: Should there be an attribute for it being allowed vs denied?
```

Comment:
@ValarDragon If we return an `Ok(response...)` it's always allowed. Otherwise we return a `Err(ContractError::...)`. Most of the times this will be `ContractError::RateLimitExceeded {}` but there could be other errors (unauthorized, channel doesn't exist, etc).

This is good because we want to stop the transaction (on the go side) when *any* error happens. Otherwise we'd have to check the response type.

Reply:
ah gotcha, thats a great design choice!